packages/driver-azure/src/driver/vm-creation-utils.ts

```diff
@@ -0,0 +1,281 @@
+  resourceGroupName: string,
+  networkInterfaceName: string,
+  networkClient: NetworkManagementClient
+) => networkClient.networkInterfaces.get(
+  resourceGroupName,
+  networkInterfaceName
+)
+
+export const createVirtualMachine = async (tags: Resource['tags'], nicId: string, imageRef: ImageReference, osDiskName: string, vmName: string, location: string, resourceGroupName: string, computeClient: ComputeManagementClient, sshPublicKey: string, networkSecurityGroupId: string, vmSize = 'Standard_B2s', adminUsername = 'preevy') => {
+  const vmParameters: VirtualMachine = {
+    location,
+    tags: { ...tags, [AzureCustomTags.VM_ID]: resourceGroupName },
+    osProfile: {
+      computerName: vmName,
+      adminUsername,
+      linuxConfiguration: {
+        ssh: {
+          publicKeys: [
+            {
+              path: `/home/${adminUsername}/.ssh/authorized_keys`,
+              keyData: sshPublicKey,
+            },
+          ],
+        },
+      },
+    },
+    hardwareProfile: {
+      vmSize,
+    },
+    storageProfile: {
+      imageReference: imageRef,
+      osDisk: {
+        name: osDiskName,
+        caching: 'None',
+        createOption: 'fromImage',
+        diskSizeGB: 32,
+        managedDisk: {
+          storageAccountType: 'StandardSSD_LRS',
+        },
+      },
+    },
+    networkProfile: {
+      networkInterfaces: [
+        {
+          id: nicId,
+          primary: true,
+        },
+      ],
+    },
+  }
+  return computeClient.virtualMachines.beginCreateOrUpdateAndWait(
+    resourceGroupName,
+    vmName,
+    vmParameters
+  )
+}
+export const extractResourceNameFromId = (rId: string) => {
+  const name = rId.split('/')
+    .at(-1)
+  if (!name) {
+    throw new Error(`Could not extract resource name from id ${rId}`)
+  }
+  return name.toLowerCase()
+}
+export const extractResourceGroupNameFromId = (rId: string) => {
+  const resourceGroupName = rId.split('/')[4]
+  if (!resourceGroupName.length) {
+    throw new Error(`Could not extract resource group name from id ${rId}`)
+  }
+  return resourceGroupName.toLowerCase()
+}
+export const getIpAddresses = async (networkClient: NetworkManagementClient, vm: VirtualMachine) => {
+  if (!vm.id || !vm.networkProfile?.networkInterfaces?.[0].id) {
+    throw new Error('Network interface configuration not found')
+  }
+  const resourceGroupName = extractResourceGroupNameFromId(vm.id)
+  const nicName = extractResourceNameFromId(vm.networkProfile.networkInterfaces[0].id)
+  const nic = await getNicInfo(resourceGroupName, nicName, networkClient)
+  if (!nic.ipConfigurations?.[0].publicIPAddress?.id) {
+    throw new Error('publicIPAddress configuration not found')
+  }
+  const publicIPName = extractResourceNameFromId(nic.ipConfigurations[0].publicIPAddress.id)
+  const publicIPAddress = await networkClient.publicIPAddresses.get(resourceGroupName, publicIPName)
+  if (!publicIPAddress.ipAddress || !nic.ipConfigurations?.[0].privateIPAddress) {
```

Comment:
consider using `ensureDefined`/`extractDefined` (it's in the aws driver but it can be moved to core)